applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -0,0 +1,786 @@
+{
+	int ret = 0;
+
+	if (data != NULL) {
+		ret = lwm2m_carrier_app_data_send(data, len);
+		if (ret) {
+			(void)exit_datamode(-EAGAIN);
+		}
+	}
+
+	return ret;
+}
+
+/* AT#XCARRIER="app_data"[,<data>] */
+static int do_carrier_appdata_send(void)
+{
+	int ret;
+
+	if (at_params_valid_count_get(&at_param_list) == 2) {
+		/* enter data mode */
+		ret = enter_datamode(carrier_datamode_callback);
+		if (ret) {
+			return ret;
+		}
+		carrier_data_mode_handler = carrier_app_data_send_handler;
+	} else {
+		char data[CONFIG_SLM_CARRIER_APP_DATA_CONTAINER_BUFFER_LEN] = {0};
+		int size = CONFIG_SLM_CARRIER_APP_DATA_CONTAINER_BUFFER_LEN;
+
+		ret = util_string_get(&at_param_list, 2, data, &size);
+		if (ret) {
+			return ret;
+		}
+
+		ret = lwm2m_carrier_app_data_send(data, size);
+	}
+
+	return ret;
+}
+
+/* AT#XCARRIER="battery_level",<battery_level> */
+static int do_carrier_device_battery_level(void)
+{
+	int ret;
+	uint16_t battery_level;
+
+	ret = at_params_unsigned_short_get(&at_param_list, 2, &battery_level);
+	if (ret) {
+		return ret;
+	}
+
+	return lwm2m_carrier_battery_level_set(battery_level);
+}
+
+/* AT#XCARRIER="battery_status",<battery_status> */
+static int do_carrier_device_battery_status(void)
+{
+	int ret, battery_status;
+
+	ret = at_params_int_get(&at_param_list, 2, &battery_status);
+	if (ret) {
+		return ret;
+	}
+
+	return lwm2m_carrier_battery_status_set(battery_status);
+}
+
+/* AT#XCARRIER="current",<power_source>,<current> */
+static int do_carrier_device_current(void)
+{
+	int ret, current;
+	uint16_t power_source;
+
+	ret = at_params_unsigned_short_get(&at_param_list, 2, &power_source);
+	if (ret) {
+		return ret;
+	}
+
+	ret = at_params_int_get(&at_param_list, 3, &current);
+	if (ret) {
+		return ret;
+	}
+
+	return lwm2m_carrier_power_source_current_set((uint8_t)power_source, current);
+}
+
+/* AT#XCARRIER="error","add|remove",<error> */
+static int do_carrier_device_error(void)
+{
+	int ret;
+	int32_t error_code;
+	char operation[7];
+	int size = sizeof(operation);
+
+	ret = util_string_get(&at_param_list, 2, operation, &size);
+	if (ret) {
+		return ret;
+	}
+
+	ret = at_params_int_get(&at_param_list, 3, &error_code);
+	if (ret) {
+		return ret;
+	}
+
+	if (slm_util_cmd_casecmp(operation, "ADD")) {
+		return lwm2m_carrier_error_code_add(error_code);
+	} else if (slm_util_cmd_casecmp(operation, "REMOVE")) {
+		return lwm2m_carrier_error_code_remove(error_code);
+	}
+
+	LOG_DBG("AT#XCARRIER=\"error\" failed: invalid operation");
+
+	return -EINVAL;
+}
+
+/* AT#XCARRIER="memory_free" */
+static int do_carrier_device_mem_free(void)
+{
+	int memory_free;
+
+	memory_free = lwm2m_carrier_memory_free_read();
+
+	sprintf(rsp_buf, "\r\n#XCARRIER: %d\r\n", memory_free);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	return 0;
+}
+
+/* AT#XCARRIER="memory_total",<memory> */
+static int do_carrier_device_mem_total(void)
+{
+	int ret, memory_total;
+
+	ret = at_params_int_get(&at_param_list, 2, &memory_total);
```

Comment:
I wonder if memory_total points to a string here, not an int? I did not get this to work. If it does you may need to take a closer look at other places where you do writes

Reply:
It does point to an int. I had no problems getting this command to work as intended. What value did you try to write here?

Reply:
I got it working. I was running a link-down link-up test right before memory total. That caused issues with the test server.